fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerTest.java

```diff
@@ -2147,6 +2150,157 @@ public void run() {
+        // failures until it hits 1
+        for (int i = 100; i > 1; i = (3 * i) / 4) {
+            queue.add(Pair.of(i, createException));
+        }
+        // queue size = 13
+        // success for a while
+        for (int i = 0; i < 10; i++) {
+            queue.add(Pair.of(1, null));
+        }
+        // queue size = 23
+        // now starts re-increasing
+        queue.add(Pair.of(2, null));
+        queue.add(Pair.of(3, null));
+        queue.add(Pair.of(4, null));
+        for (int i = 5; i < 100; i = (i * 4) / 3) {
+            queue.add(Pair.of(i, null));
+        }
+        // queue size = 38
+        // does not pass original max
+        queue.add(Pair.of(100, null));
+        queue.add(Pair.of(100, null));
+        queue.add(Pair.of(100, null));
+        for (int i = 100; i > 42; i = (3 * i) / 4) {
+            queue.add(Pair.of(i, createException));
+        }
+        // queue size = 44
+        // success for a while
+        for (int i = 0; i < 10; i++) {
+            queue.add(Pair.of(42, null));
+        }
+        // queue size = 54
+        // fail once
+        queue.add(Pair.of(56, createException));
+        for (int i = 0; i < 10; i++) {
+            queue.add(Pair.of(42, null));
+        }
+        // queue size = 65
+        queue.add(Pair.of(56, createException));
+        queue.add(Pair.of(42, null));
+
+        reincreaseLimit(queue, index ->
+                OnlineIndexer.newBuilder()
+                        .setDatabase(fdb).setMetaData(metaData).setIndex(index).setSubspace(subspace)
+                        .setLimit(100).setMaxRetries(queue.size() + 3).setRecordsPerSecond(10000)
+                        .setIncreaseLimitAfter(10)
+                        .setMdcContext(ImmutableMap.of("mdcKey", "my cool mdc value"))
+                        .setMaxAttempts(3)
+                        .build());
+    }
+
+    private void reincreaseLimit(Queue<Pair<Integer, Supplier<RuntimeException>>> queue,
+                                 final Function<Index, OnlineIndexer> buildOnlineIndexer) {
+        Index index = new Index("newIndex", field("num_value_2"));
+        openSimpleMetaData(metaDataBuilder -> metaDataBuilder.addIndex("MySimpleRecord", index));
+
+        try (FDBRecordContext context = openContext()) {
+            // OnlineIndexer.runAsync checks that the index is not readable
+            recordStore.clearAndMarkIndexWriteOnly(index).join();
+            context.commit();
+        }
+        try (OnlineIndexer indexBuilder = buildOnlineIndexer.apply(index)) {
+
+            AtomicInteger attempts = new AtomicInteger();
+            attempts.set(0);
+            AsyncUtil.whileTrue(() ->
+                    indexBuilder.runAsync(store -> {
+                        Pair<Integer, Supplier<RuntimeException>> behavior = queue.poll();
+                        if (behavior == null) {
+                            return CompletableFuture.completedFuture(false);
```

Comment:
Could be `AsyncUtil.READY_FALSE` (and the other line can be `AsyncUtil.READY_TRUE`), though maybe it doesn't matter.